src/cupbearer/detectors/statistical/statistical.py

```diff
@@ -90,6 +103,42 @@ def batch_update(self, activations: dict[str, torch.Tensor]):
     def post_covariance_training(self, **kwargs):
         pass
 
+    @abstractmethod
+    def _individual_layerwise_score(self, name: str, activation: torch.Tensor):
+        """Compute the anomaly score for a single layer.
+
+        `name` is passed in to access the mean/covariance and any custom derived
+        quantities computed in post_covariance_training.
+
+        `activation` will always have shape (batch, dim). The `batch` dimension might
+        not just be the actual batch dimension, but could also contain multiple entries
+        from a single sample, in the case of multi-dimensional activations that we
+        treat as independent along all but the last dimension.
+
+        Should return a tensor of shape (batch,) with the anomaly scores.
+        """
+        pass
+
+    def layerwise_scores(self, batch):
+        activations = self.get_activations(batch)
+        batch_size = next(iter(activations.values())).shape[0]
+        activations = {
+            k: rearrange(v, "batch ... dim -> (batch ...) dim")
```

Comment:
So `v.flatten(end_dim=-2)`? I'm a bit unused to this notation.

Reply:
Yes, same thing. Personally, I really like einops (though this case is simple enough that either seems fine)